spec/models/accounting/quickbooks/transaction_creator_spec.rb

```diff
@@ -2,28 +2,109 @@
-    expect(generic_service).to receive(:create)
-    expect(class_service).to receive(:find_by)
-
-    subject.add_disbursement(
-      amount: 12.09,
-      loan_id: 2,
-      memo: 'I am memo',
-      description: 'I am a line item description',
-      qb_bank_account_id: 89,
-      qb_customer_id: 3
-    )
+  shared_examples_for 'JournalEntry service' do
+    it 'calls create with correct data' do
+      expect(generic_service).to receive(:create) do |arg|
+        expect(arg.line_items.count).to eq 2
+        expect(arg.private_note).to eq memo
+        expect(arg.txn_date).to be_nil
+
+        list = arg.line_items
+        expect(list.map(&:amount).uniq).to eq [amount]
+        expect(list.map(&:description).uniq).to eq [description]
+
+        details = list.map { |i| i.journal_entry_line_detail }
+        expect(details.map { |i| i.posting_type }.uniq).to match_array %w(Debit Credit)
+        expect(details.map { |i| i.entity.type }.uniq).to eq %w(Customer)
+        expect(details.map { |i| i.entity.entity_ref.value }.uniq).to eq [qb_customer_id]
+        expect(details.map { |i| i.class_ref.value }.uniq).to eq [loan_id]
+        expect(details.map { |i| i.account_ref.value }.uniq).to match_array [qb_bank_account_id, qb_principal_account_id]
+      end
+      subject
+    end
+
+    context 'and date is supplied' do
+      let(:date) { 3.days.ago }
+
+      it 'creates JournalEntry with date' do
+        expect(generic_service).to receive(:create) do |arg|
+          expect(arg.txn_date).to eq date
+        end
+        subject
+      end
+    end
+
+    it 'creates JournalEntry with a reference to the existing loan' do
+      expect(class_service).to receive(:find_by).with(:name, loan_id)
+      subject
+    end
+  end
+
+  context 'when qb customer does not exist' do
+    let(:qb_customer_id) { '91234' }
+    let(:qb_new_customer) { instance_double(Quickbooks::Model::Customer, id: qb_customer_id) }
+    let(:customer_service) { instance_double(Quickbooks::Service::Customer, create: qb_new_customer) }
+    let(:customer_name) { 'A cooperative with a name' }
+    let(:organization) { create(:organization, name: customer_name, qb_id: nil) }
+
+    it_behaves_like 'JournalEntry service'
+
+    it 'creates Customer with using name' do
+      expect(customer_service).to receive(:create) do |arg|
+        expect(arg.display_name).to eq customer_name
+      end.and_return(qb_new_customer)
+      subject
+    end
+  end
+
+  context 'when qb customer does exist' do
+    let(:qb_customer_id) { '3' }
+    let(:organization) { create(:organization, qb_id: qb_customer_id) }
+
+    it_behaves_like 'JournalEntry service'
+
+    context 'and date is supplied' do
+      let(:date) { 3.days.ago }
+
+      it 'creates JournalEntry with date' do
+        expect(generic_service).to receive(:create) do |arg|
+          expect(arg.txn_date).to eq date
+        end
+        subject
+      end
+    end
```

Comment:
The above two contexts should move into a new Customer spec. Then this spec can just stub the customer fetching stuff and assume it works properly. :)